### PR TITLE
fix: devspace deploy do git pull for cached dependencies

### LINKED ISSUE #2765: Devspace deploy do not git pull for cached dependencies
**What happened?**  
As part of this PR https://github.com/devspace-sh/devspace/pull/2709 i introduced a regression bug where it now never attempts to `git pull` for cached dependencies.

I have created a PR https://github.com/devspace-sh/devspace/pull/2766 to revert this with some refactoring.

**What did you expect to happen instead?**  
Given the configuration

``` yaml
dependencies:
  dependency-a:
    git: https://www.somewhere.com/dependency-a
    branch: master
    disablePull: false
```

I expect devspace deploy to pull new commits to dependency-a that has happen since the last devspace deploy.

**Local Environment:**  
- DevSpace Version: 6.3.5
- Operating System: windows 
- ARCH of the OS: AMD64
**Kubernetes Cluster:**  
- Cloud Provider: kind
- Kubernetes Version: 1.27.2

**Anything else we need to know?**  
Read https://github.com/devspace-sh/devspace/pull/2766

### PR DESCRIPTION
**Pull request by Andrioden**
**What issue type does this pull request address?**
/kind bugfix
/kind enhancement

**What does this pull request do? Which issues does it resolve?**
resolves #2765

**Please provide a short message that should be published in the DevSpace release notes**
Same as PR title

**What else do we need to know?** 
Alternatively the PR https://github.com/devspace-sh/devspace/pull/2709 could be straight up reverted. As reverting would be (i think) functionally the same as this PR. However, it does not read very well. It was not a coincidence that i misunderstood the "git clone" code path of DownloadDependency that i was attempting to fix in my last pr, thought, I am no go developer and are in that sense reading and writing code with a handicap. :P 

**Testing**
As I am not able to write integration tests, i have manually tested a locally built devspace.exe:
- It correctly git clone repos when i delete .dependencies folder
- It correctly reclone git repo when a branch is changed
- It correctly git pull cached dependencies
- It respects the dependency.disablePull flag by not pulling when its true
- It correctly tries to switch the url type (switchURLType(gitPath))

### CODE REVIEW

File: pkg/devspace/dependency/util/util.go [hunk 1]
```diff
@@ -142,8 +140,13 @@ func DownloadDependency(ctx context.Context, workingDirectory string, source *la
 					return "", errors.Wrap(err, "clone repository")
 				}
 			}
+
+			log.Debugf("Cloned %s", gitPath)
+		} else if !source.DisablePull {
+			repo.Pull(ctx, gitCloneOptions)
```
Comment (lizardruss) [inline pkg/devspace/dependency/util/util.go:146]: If pulling is removed from `repo.Clone`, then we could change this from an `else if` to an `if`:
```golang
if !source.DisablePull && source.Revision == "" {
  // Pull
}
```

May need to also handle the error, and attempt to switching the URL, similar to [this line](https://github.com/devspace-sh/devspace/pull/2766/files#diff-a5251b488fef0b4fa9e68c8210ca97fc0553bd1507255378cb8cee3ef079fd53R129).

**Review by lizardruss — CHANGES_REQUESTED** [review]
Thanks for the PR! I like where this is going.

I left some comments, but they may not be clear. More generally, I'd like to call `repo.Clone()` and `repo.Pull()` separately, so that it becomes clearer that we skip pulling when a revision has been specified.

I think this would also get rid of the if / else between clone and pull, and make it more like
- clone (if needed)
- pull (if no revision is specified and pulling is not disabled)

This might also make it possible to push the `switchURL()` logic into `repo.Clone` & `repo.Pull`, but I'm ok with duplicating it for logging purposes.

**Comment by Andrioden** [comment]
Will look into comments later this week

**Comment by Andrioden** [comment]
ℹ️ Output from newest testing/code, looks good (with considerations within the scope of this PR)

```
14:35:39 debug Cloned https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Configuration
14:35:41 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Configuration
14:35:42 debug Cloned https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Oppsett
14:35:43 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Oppsett
14:35:45 debug Cloned https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-FellesKonfig
14:35:46 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-FellesKonfig
14:35:48 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Oppsett
14:35:50 debug Cloned https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Felleslogg
14:35:51 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Felleslogg
14:35:52 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Oppsett
14:35:58 debug Cloned https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Personvern
14:35:59 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Personvern
14:36:01 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Oppsett
14:36:07 debug Cloned https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-STS
14:36:08 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-STS
14:36:09 debug Pulled https://nhnfelles@dev.azure.com/nhnfelles/Helsenorge/_git/HN-Oppsett
```

File: pkg/devspace/dependency/util/util.go (hunk 1 above)
Comment (Andrioden) [reply pkg/devspace/dependency/util/util.go]: I agree, take a look again.

I am not sure if me messing up the commit history matters? I hope we can squash into the origin repository.

I retested
- It correctly git clone repos when i delete .dependencies folder
- I correctly reclone git repo when a branch is changed
- It correctly git pull cached dependencies
- It respects the dependency.disablePull flag by not pulling when its true

File: pkg/util/git/git.go [hunk 2]
```diff
@@ -87,7 +87,13 @@ func (gr *GitCLIRepository) Clone(ctx context.Context, options CloneOptions) err
 		return nil
 	}
 
-	// make sure the repo is up-to-date
+	// Make sure the repo is up-to-date
+	gr.Pull(ctx, options)
```
[the remark this answers is not in the document]
Comment (Andrioden) [reply pkg/util/git/git.go]: Agreed

**Comment by Andrioden** [comment]
I believe this is ready for merge @lizardruss

**Review by lizardruss — APPROVED** [review]
Thanks for the changes! LGTM! Will run tests and squash the commits when merging.

**Comment by Andrioden** [comment]
@lizardruss - Fixed lint errors, at least i think i have, i was not able to reproduce them locally, and the workflow here is not run until you trigger them.

**Comment by Andrioden** [comment]
@lizardruss - Any chance this could be merged?

**Pull request merged**